### ISSUE TITLE
Using not yet defined variables is allowed

### ISSUE THREAD

**Issue by afrieder** [python/mypy#2788]
This might be the same issue as #937 but this raises no errors:

```python
s += 1
s: int = 0
```

**Comment by JukkaL** [comment]
Mypy will catch the first case if it's within a function:

```
def f() -> None:
    s += 1   # Name 's' is not defined
    s: int = 0
```

This should be rejected at the top level of a file as well. The second case is arguably different and would require keeping track of when variables are defined or not. I remember having seen an issue about this but I can't find it right now.

**Comment by KotlinIsland** [comment]
What you could do is assign unset names with a type of `X | Unbound`, I think Pyright does something like this:
```py
a: int
reveal_type(a)  # int | Unbound
print(a + 1)  # error, usage of Unbound 
a = 1
reveal_type(a)  # int
```

**Comment by KotlinIsland** [comment]
Duplicate of #686?
